blockstore/apps/bundles/store.py

```diff
@@ -208,68 +203,128 @@ class BundleDataStore():
+    def __init__(self, storage=None):
+        self.storage = storage or default_storage
+
+    def get(self, bundle_uuid: UUID, snapshot_digest: bytes) -> Snapshot:
+        """
+        Return a snapshot.
 
-        return BundleSnapshot(
+        TODO: There's no meaningful error handling yet.
+        """
+        storage_path = self._summary_path(bundle_uuid, snapshot_digest)
+        try:
+            with self.storage.open(storage_path) as snapshot_file:
+                snapshot_json = json.load(snapshot_file)
+        except FileNotFoundError as err:
+            logger.error(
+                "Snapshot %s,%s not found at: %s",
+                bundle_uuid,
+                snapshot_digest.hex(),
+                storage_path,
+            )
+            raise Snapshot.NotFoundError(
+                f"Snapshot {snapshot_digest.hex()} for Bundle {bundle_uuid} not found"
+            )
+
+        return Snapshot(
             bundle_uuid=bundle_uuid,
-            files=files,
-            hash_digest=hash_digest,
-            created_at=created_at,
+            files=FileInfo.from_json_dict(snapshot_json['files']),
+            hash_digest=bytes_from_hex_str(snapshot_json['hash_digest']),
+            created_at=parse_utc_iso8601_datetime(snapshot_json['created_at']),
         )
 
+    def create(self, bundle_uuid, paths_to_files):
+        """
+        Save the files, create a Snapshot object and save its JSON serialization to storage.
+        """
+        files = {}
+        for path, file in paths_to_files.items():
+            files[str(path)] = self._save_file(bundle_uuid, path, file)
+
+        return self._create(bundle_uuid, files)
+
+    def create_by_adding_paths(self, snapshot, paths_to_files):
+        """
+        Create and save a Snapshot with a file added at path to snapshot.
+        """
+        files = dict(snapshot.files)
+        for fileinfo in paths_to_files:
+            files[str(fileinfo['path'])] = self._save_file(snapshot.bundle_uuid, **fileinfo)
+        return self._create(snapshot.bundle_uuid, files)
+
+    def create_by_removing_path(self, snapshot, path):
+        """
+        Create and save a Snapshot with file at path removed from snapshot.
+        """
+        files = {p: snapshot.files[p] for p in snapshot.files if p != path}
+        return self._create(snapshot.bundle_uuid, files)
+
+    def url(self, snapshot, path):
+        """Return a user-accessible URL to download a path from this Snapshot."""
+        file_info = snapshot.files[path]  # pylint: disable=unsubscriptable-object
+        storage_path = self._file_data_path(snapshot.bundle_uuid, file_info.hash_digest.hex())
+        return self.storage.url(storage_path)
+
+    def open(self, snapshot, path):
+        file_info = snapshot.files[path]  # pylint: disable=unsubscriptable-object
+        storage_path = self._file_data_path(snapshot.bundle_uuid, file_info.hash_digest.hex())
+        return self.storage.open(storage_path, 'rb')
```

Comment:
What are the circumstances under which we'd read file contents within a `SnapshotRepo`? I think we should avoid reading file contents from the repo like the plague for performance reasons (only ever returning URLs and letting clients read file content directly from S3/CloudFront). Obviously we need the hash on the initial upload, but this method is for reading the content of files that have already been uploaded into the repo.

From what I can tell, this is only used (via `file_mapping`) in `DraftRepo.commit()`, but it "feels" to me like it should be possible to commit a draft without Blockstore needing to read any file contents. We already know the hash and size of every file in the draft.

Reply:
So I was mucking with this a bit last night...

I agree that opening files should be avoided, and I like the idea of not even allowing an interface for that. The original reason I put this in is because I wanted some kind of barrier between Drafts and Snapshots, because I feel a lot more uncertainty about Drafts in the long term, and I'd like to have the flexibility to keep it more separate (different backing store, different whatever) so that it's easier to toss aside and replace. I didn't want to cram in a formal separation by app in this PR because it's already way too big and been open way too long, but I figured that later on it would end up in that direction.

On the other hand, the only way to make it so that we can guarantee that `commit()` is a fast, synchronous operation is if drafts share the same backing store and we can quickly do moves of files instead of rewriting them. But unfortunately, moves don't exist in the Django File Storage API. We'd have to dig down to lower level abstractions and have our own custom backends for AWS and other backends if we wanted to maintain that.

So is it okay if we leave this in here for now, and try to figure out the bigger story about the layer of abstraction for Drafts becoming Snapshots afterwards?

Reply:
I see. Yep, as discussed, I'm keen to get this merged so we can figure that out in a future PR.